### PR TITLE
Add min compat version option to describeCompat parameter

### PR DESCRIPTION
**Pull request by dannimad**
Removing "NoCompat"  describeCompat option and substitute that by specifying minimum compat version (string) available for a specific test suite.

Usage example:
 `describeCompat("Detached Container", "2.0.0-internal.8.0.0", (getTestObjectProvider, apis) => {`

 `describeCompat` expects 3 arguments: (name: string, compatVersion: string, tests).
 There are three compatVersion options to generate different combinations, depending of the need of the tests:
 - `FullCompat`: generate test variants with compat combinations that varies the version for all layers.
 - `LoaderCompat`: generate test variants with compat combinations that only varies the loader version.
 - Specific version (String) : specify a minimum compat version (e.g. "2.0.0-rc.1.0.0") which will be the minimum version a test suite will test against. This should be equal to the value of pkgVersion at the time you're writing the new test suite.

### CODE REVIEW

File: packages/test/test-version-utils/src/versionUtils.ts [hunk 1]
```diff
@@ -160,6 +160,22 @@ export function resolveVersion(requested: string, installed: boolean) {
 	}
 }
 
+/**
+ * @internal
+ */
+export function getAllFluidVersions(): Map<string, number> {
+	const allVersionsFromNpm = execSync(`npm show fluid-framework versions --json`, {
```
Comment (dannimad) [inline packages/test/test-version-utils/src/versionUtils.ts:167]: is there a better/faster way to get fluid versions?

File: packages/test/test-version-utils/src/describeCompat.ts [hunk 2]
```diff
@@ -38,14 +39,19 @@ function createCompatSuite(
...
 		let configs = configList.value;
 		if (compatFilter !== undefined) {
 			configs = configs.filter((value) => compatFilter.includes(value.kind));
 		}
-
+		const versionsMap = getAllFluidVersions();
 		for (const config of configs) {
+			if (minVersion && isCompatVersionBelowMinVersion(minVersion, versionsMap, config)) {
+				// skip current config if compat version is below min version supported for test suite
+				continue;
```
Comment (dannimad) [inline packages/test/test-version-utils/src/describeCompat.ts:53]: There is an optimization we could make here given that we are filtering tons of configs, here and lines above with compatFilter. mochaGlobalSetup ensures all versions got from the configList are installed when there is no need for some of the filtered configs. Changes could be considerable so maybe it could be made in a follow up PR/ task

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { CompatKind } from "../../compatOptions.cjs";
+import { getAllFluidVersions } from "../versionUtils.js";
+import { pkgVersion } from "../packageVersion.js";
+import { isCompatVersionBelowMinVersion } from "../compatConfig.js";
+
+describe("Minimum Compat Version", () => {
```
Comment (dannimad) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:12]: ideas for more test coverage are welcome

File: packages/test/test-version-utils/src/versionUtils.ts [hunk 4]
```diff
@@ -360,7 +376,7 @@ export function getRequestedVersion(
 		return baseVersion;
 	}
 	if (typeof requested === "string") {
-		return requested;
+		return resolveVersion(requested, false);
```
Comment (dannimad) [inline packages/test/test-version-utils/src/versionUtils.ts:379]: by using resolveVersion, strings in formats like "^1.3.4" are normalized to "1.3.7" which is the version that is gonna be used anyway.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 5]
```diff
@@ -238,6 +266,7 @@ export const genCrossVersionCompatConfig = (): CompatConfig[] => {
 						compatVersion: resolvedCreateVersion,
 						createWith: createVersion,
 						loadWith: loadVersion,
+						loadVersion: resolvedLoadVersion,
```
Comment (dannimad) [inline packages/test/test-version-utils/src/compatConfig.ts:269]: We need both versions (load and create) to compare the minimum of those against the specified minVersion.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 6]
```diff
@@ -198,6 +200,32 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
 	return _configList;
 };
 
+export function isCompatVersionBelowMinVersion(
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:203]: Can you add a description to this function and comments inside the function. It's very hard to follow what is happening.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 7]
```diff
@@ -198,6 +200,32 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
+	minVersion: string,
+	allVersions: string[],
+	config: CompatConfig,
+) {
+	let lowerVersion: string | number;
+	if (config.kind === CompatKind.CrossVersion) {
+		const compatV = allVersions.indexOf(config.compatVersion as string);
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:210]: Not sure I understand why allVersions is needed. Why not semver compare the 2 versions here?

File: packages/test/test-version-utils/src/describeCompat.ts [hunk 8]
```diff
@@ -221,37 +227,36 @@ export type DescribeCompat = DescribeCompatSuite &
...
 	const createCompatSuiteWithDefault = (
 		tests: (this: Mocha.Suite, provider: () => ITestObjectProvider, apis: CompatApis) => void,
-		compatVersionKind: CompatVersionKind,
+		compatVersion: string,
 	) => {
-		switch (compatVersionKind) {
+		switch (compatVersion) {
 			case "FullCompat":
 				return createCompatSuite(tests, undefined);
 			case "NoCompat":
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/describeCompat.ts:235]: I think "NoCompat" should go away with this change. It should either be FullCompat or specify a min version.

File: packages/test/test-version-utils/src/compatConfig.ts (hunk 7 above)
Comment (dannimad) [reply packages/test/test-version-utils/src/compatConfig.ts]: I just push a commit with that approach

File: packages/test/test-version-utils/src/describeCompat.ts (hunk 8 above)
Comment (dannimad) [reply packages/test/test-version-utils/src/describeCompat.ts:234]: There are several issues with removing NoCompat:
1. All current NoCompat test would need to be changed to specify a min version. It will be very hard to know which version each suite was intended to be tested against in the first place.
2. Even if we do the job and specify a min version to all current NoCompat tests, curretn infrastructure won't support such a change. Scott and I have tried to change current NoCompat tests that already work FullCompat and pipeline timeout is almost guaranteed. 
3. Current rc version and 2.0.0-internal.8.0.0 are not getting correct N-i versions. (don't know if it is because current special rc version).
4. I would wait for the next release to tackle some of the issues described before and only once they are fixed, deprecate NoCompat.

Comment (scottn12) [reply packages/test/test-version-utils/src/describeCompat.ts:234]: > Current rc version and 2.0.0-internal.8.0.0 are not getting correct N-i versions. (don't know if it is because current special rc version).

FYI, this should be fixed after https://github.com/microsoft/FluidFramework/pull/19092 / https://github.com/microsoft/FluidFramework/pull/19127 🙂

Comment (dannimad) [reply packages/test/test-version-utils/src/describeCompat.ts:234]: Even if some of the issues are already fixed, I think we still need to keep NoCompat and slowly remove those to either a min version or full compat.

Comment (agarwal-navin) [reply packages/test/test-version-utils/src/describeCompat.ts:234]: It's okay if that doesn't happen in this PR. You can create a follow up task to tackle this.

> All current NoCompat test would need to be changed to specify a min version. It will be very hard to know which version each suite was intended to be tested against in the first place.

I wouldn't bother finding out when the test was added but use a recent version - maybe internal-7.0 or something. Eventually, these tests will start testing fullCompat.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 9]
```diff
@@ -206,6 +209,30 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
 	return _configList;
 };
 
+/**
+ * Returns true if compat test version is below the one provided as minimum version.
+ * It helps to filter lower verions configs that the ones intended to be tested on a
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:214]:
```suggestion
 * It helps to filter out lower verions configs than the ones intended to be tested on a
```

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 10]
```diff
@@ -206,6 +209,30 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
+ * particular suite.
+ * @param minVersion -
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:216]: These param descriptions are empty. You should either add a description or remove them.

File: packages/test/test-version-utils/src/compatConfig.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/test/test-version-utils/src/compatConfig.ts:276]: You can avoid adding `loadVersion` by generating the `resolvedLoadVersion` from `loadWith` in `isCompatVersionBelowMinVersion` as it's only used there.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 11]
```diff
@@ -52,6 +53,7 @@ export interface CompatConfig {
 	 * (Same version will be used across all layers).
 	 */
 	loadWith?: CompatVersion;
+	loadVersion?: string;
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:56]: As per my other comment you can remove this and generate this from `loadWith`. If you decide to keep it, add a comment here similar to the ones for createWith and loadWith which says this is for cross version compat only and that this is basically a resolved version from loadWith.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 12]
```diff
@@ -206,6 +209,30 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
+/**
+ * Returns true if compat test version is below the one provided as minimum version.
+ * It helps to filter lower verions configs that the ones intended to be tested on a
+ * particular suite.
+ * @param minVersion -
+ * @param config -
+ * @returns boolean
+ */
+export function isCompatVersionBelowMinVersion(minVersion: string, config: CompatConfig) {
+	let lowerVersion: string | number;
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:221]: nit: You can do `let lowerVersion = config.compatVersion;` here. That way you don't need the else block below. Just a small optimization.

File: packages/test/test-version-utils/src/describeCompat.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/test/test-version-utils/src/describeCompat.ts:234]: One desirable outcome of this PR (and any immediate follow-up) would be that new tests are not added with "NoCompat". With that in mind, do you think we should deprecate "NoCompat" option so that anyone adding it immediately knows that there is an alternative? Although for that you will have to keep `CompatVersionKind` and maybe add the `minVersion` as a separate parameter. Maybe that's okay?

File: packages/test/test-version-utils/src/describeCompat.ts [hunk 13]
```diff
@@ -204,7 +209,7 @@ export interface ITestObjectProviderOptions {
  */
 export type DescribeCompatSuite = (
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/describeCompat.ts:210]: Can you update an existing end-to-end test (or add a new one) that uses this new pattern. It would be helpful to see it in action and review it from that perspective as well.

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+import { execSync } from "child_process";
+import { CompatKind } from "../../compatOptions.cjs";
+import { isCompatVersionBelowMinVersion } from "../compatConfig.js";
+
+describe("Minimum Compat Version", () => {
+	const allVersionsFromNpm = execSync(`npm show fluid-framework versions --json`, {
+		encoding: "utf-8",
+	});
+	const allVersions: string[] = JSON.parse(allVersionsFromNpm);
+	const latestVersion = allVersions[allVersions.length - 1];
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:16]: Is this really needed? Can't you just use `pkgVersion`?

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+	it("bad min compat string", () => {
+		const invalidString = "invalid string";
+		try {
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:20]: You can use `assert.throws` instead of this try-catch block which is a lot cleaner.

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+					compatVersion: -i,
+				}),
+				true,
+				`N-${i} is not lower than min version`,
+			);
+		});
+	}
+
+	it("cross compat", () => {
+		assert.strictEqual(
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:50]: nit: These 4 different asserts may be better as separate test cases as they are testing different aspects of the API. At least you should add a comment / fail message to each that will help with debugging / understanding.

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+		} catch (error: any) {
+			assert.strictEqual(
+				error.message,
+				`Error while running: npm v @fluidframework/container-loader@"${invalidString}" version --json`,
+			);
+		}
+	});
+
+	for (let i = 1; i < 9; i++) {
+		it(`compatVersion N-${i} < latest version ${latestVersion}`, () => {
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:36]: Maybe add couple of cases where compat version is not below min version.

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+			assert.strictEqual(
+				isCompatVersionBelowMinVersion(latestVersion, {
+					name: `test`,
+					kind: CompatKind.None,
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:40]: Does it make sense to add some test with different compat kinds?

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+		});
+	}
+
+	it("cross compat", () => {
+		assert.strictEqual(
+			isCompatVersionBelowMinVersion(latestVersion, {
+				name: "test",
+				kind: CompatKind.CrossVersion,
+				compatVersion: latestVersion,
+				loadVersion: "1.3.7",
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:55]: Curious as to why 1.3.7? And not like 1.0.0?

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts (hunk 18 above)
Comment (dannimad) [reply packages/test/test-version-utils/src/test/minCompatVersion.spec.ts:40]: I think it doesn't add up. The only interesting kind is CrossVersion I believe.

File: packages/test/test-version-utils/src/test/minCompatVersion.spec.ts (hunk 14 above)
Comment (dannimad) [reply packages/test/test-version-utils/src/test/minCompatVersion.spec.ts]: I tried using pkgVersion but it broke these tests when running in the pipeline. pkgVersion in ado is different than the one in repo for some reason. Trying with codeVersion but not 100% confident with that approach either

File: packages/test/test-end-to-end-tests/src/test/SummarizeFetchValidation.spec.ts [hunk 20]
```diff
@@ -110,7 +110,7 @@ async function createSummarizer(
  */
 describeCompat(
 	"Summarizer fetches expected number of times",
-	"NoCompat",
+	"2.0.0-rc.1.0.0",
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/SummarizeFetchValidation.spec.ts:113]:
```suggestion
	"2.0.0-rc.1.0.0" /* minVersion */,
```
nit: Add this and a comment explaining what it is. If we do this to the first few tests, it will carried over to other tests and it will help others understand and use it correctly.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 21]
```diff
@@ -206,6 +213,25 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
+	let lowerVersion: string | number = config.compatVersion;
+	// For CrossVersion there are 2 versions being tested. Get the lower one.
+	if (config.kind === CompatKind.CrossVersion) {
+		lowerVersion =
+			semver.compare(config.compatVersion as string, config.loadVersion as string) > 0
+				? (config.loadVersion as string)
+				: config.compatVersion;
+	}
+	const compatVersion = getRequestedVersion(testBaseVersion(lowerVersion), lowerVersion);
+	const minReqVersion = getRequestedVersion(testBaseVersion(minVersion), minVersion);
```
Comment (agarwal-navin) [inline packages/test/test-version-utils/src/compatConfig.ts:231]: I remember I looked at this and commented about it. But this will throw if the minVersion is not in correct format right? Just want to make sure that if someone uses an incorrect version, they will become aware and can fix it.

**Review by agarwal-navin — APPROVED** [review]

File: packages/test/test-version-utils/src/compatConfig.ts (hunk 21 above)
Comment (dannimad) [reply packages/test/test-version-utils/src/compatConfig.ts:231]: Yes, from your comment, I moved the error handling deeper into getRequestedVersion.

**Review by Abe27342 — APPROVED** [review]
LGTM.. do we have any follow-up work to see if we can move the compat matrix for some of these to versions before 2.0.0-rc.1.0.0?
